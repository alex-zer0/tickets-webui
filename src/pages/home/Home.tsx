import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Dispatch } from 'redux';
import { fetchEvents } from 'src/store/actions/events';
import { EventShortInfo, EventsState } from 'src/store/reducers/events';
import './Home.scss';

export interface HomeProps {
  events: EventShortInfo[] | null
  loading: boolean
  error?: string
  fetchEvents: () => void
}

export class Home extends Component<HomeProps> {
  public componentDidMount() {
    this.props.fetchEvents();
  }

  public renderError(error: string) {
    return (
      <div className='error'>
        <h2>{error}</h2>
        <p>Please, try again a little bit later</p>
      </div>
    );
  }

  public renderLoader() {
    return (
      <div className='loader'>
        <h2>Loading...</h2>
        <p>Wait a second...</p>
      </div>
    );
  }

  public render() {
    const { events, loading, error } = this.props;
    if (loading) {
      return this.renderLoader();
    }
    if (error) {
      return this.renderError(error);
    }
    if (!events) {
      return null;
    }
    return (
      <div className='Home'>
        <ul>
          {events.map((event, i) => <li key={i}><Link to={event.url}>{event.name}</Link></li>)}
        </ul>
      </div>
    );
  }
}

function mapStateToProps(state: { eventsState: EventsState }) {
  return {
    events: state.eventsState.events,
    loading: state.eventsState.loading,
  };
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    fetchEvents: () => dispatch(fetchEvents()),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Home);
