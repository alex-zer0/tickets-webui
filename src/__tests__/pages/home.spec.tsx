import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Home, HomeProps } from '../../pages/home/Home';

const homeProps = {
  events: null,
  fetchEvents: jest.fn(),
  loading: false,
};

describe('PAGES: Home', () => {
  it('renders Loader', () => {
    const props: HomeProps = {
      ...homeProps,
      loading: true,
    };
    const tree = renderer
      .create(<Home {...props}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Error', () => {
    const props: HomeProps = {
      ...homeProps,
      error: 'Test Error Message',
    };
    const tree = renderer
      .create(<Home {...props}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders as null', () => {
    const tree = renderer
      .create(<Home {...homeProps}/>)
      .toJSON();
    expect(tree).toEqual(null);
  });

  it('renders w/o exceptions', () => {
    const props: HomeProps = {
      ...homeProps,
      events: [{ name: 'Event name', url: '/event' }],
    };
    const tree = renderer
      .create(<BrowserRouter><Home {...props}/></BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
