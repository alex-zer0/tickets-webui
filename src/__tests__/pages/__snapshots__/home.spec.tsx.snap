// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAGES: Home renders Error 1`] = `
<div
  className="error"
>
  <h2>
    Test Error Message
  </h2>
  <p>
    Please, try again a little bit later
  </p>
</div>
`;

exports[`PAGES: Home renders Loader 1`] = `
<div
  className="loader"
>
  <h2>
    Loading...
  </h2>
  <p>
    Wait a second...
  </p>
</div>
`;

exports[`PAGES: Home renders w/o exceptions 1`] = `
<div
  className="Home"
>
  <ul>
    <li>
      <a
        href="/event"
        onClick={[Function]}
      >
        Event name
      </a>
    </li>
  </ul>
</div>
`;
